### PR TITLE
[Monitor]az monitor app-insights component

### LINKED ISSUE #1154: How to set daily cap?
is there a cli command or param for this? https://github.com/Azure/azure-rest-api-specs/blob/master/specification/applicationinsights/resource-manager/Microsoft.Insights/stable/2015-05-01/componentFeaturesAndPricing_API.json#L37

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 333d37d2-61e2-0184-d253-157ee236439c
* Version Independent ID: 9b0e976f-3cb5-e813-9f33-e9cea7a8df8c
* Content: [az monitor app-insights component](https://docs.microsoft.com/en-us/cli/azure/ext/application-insights/monitor/app-insights/component?view=azure-cli-latest#ext-application-insights-az-monitor-app-insights-component-create)
* Content Source: [latest/docs-ref-autogen/ext/application-insights/monitor/app-insights/component.yml](https://github.com/MicrosoftDocs/azure-docs-cli/blob/live/latest/docs-ref-autogen/ext/application-insights/monitor/app-insights/component.yml)
* GitHub Login: @rloutlaw
* Microsoft Alias: **routlaw**

### LINKED ISSUE #1269: az monitor app-insights component show --subscription is choosing a different subscription
### Extension name (the extension in question)

application-insights

### Description of issue (in as much detail as possible)

When I run this:

```ps
$vSubscription = "abc"
az monitor app-insights component show --subscription $vSubscription
```

It returns results for a completely different subscription. I can give more details about the subscriptions in question if you reach out on Teams.

This is blocking me from using the extension to pull the instrumentation key via the CLI.

### PR DESCRIPTION
**Pull request by MyronFanQiu**
Fix #1154 support setting cap
Fix #1269 fix for --subscription

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### CODE REVIEW

File: src/application-insights/azext_applicationinsights/commands.py [hunk 1]
```diff
@@ -60,6 +70,10 @@ def load_command_table(self, _):
         g.custom_command('delete', 'delete_component')
         g.custom_command('update-tags', 'update_component_tags')
 
+    with self.command_group('monitor app-insights component billing', command_type=components_billing_sdk, custom_command_type=components_billing_custom_sdk) as g:
+        g.custom_command('update', 'update_componet_billing')
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/commands.py]: Do we need generic update to support 'patch' as it seems SDK uses 'put'?

File: src/application-insights/azext_applicationinsights/_help.py [hunk 2]
```diff
@@ -88,6 +88,21 @@
...
 """
 
+helps['monitor app-insights component billing'] = """
+    type: group
+    short-summary: Manage an Application Insights component billing features.
+"""
+
+helps['monitor app-insights component billing show'] = """
+    type: command
+    short-summary: Show the billing features of an Application Insights resource.
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/_help.py]: No example?

File: src/application-insights/azext_applicationinsights/commands.py (hunk 1 above)
Comment (MyronFanQiu) [reply src/application-insights/azext_applicationinsights/commands.py]: good catch. I will make it as patch-like. However, I don't think we need to use generic update. The model contains about six fields. But I can only update two of them. I think exposing the other arguments to the customers may cause issues.

File: src/application-insights/azext_applicationinsights/_help.py (hunk 2 above)
Comment (MyronFanQiu) [reply src/application-insights/azext_applicationinsights/_help.py]: I can add one. But it's quite simple.

File: src/application-insights/azext_applicationinsights/_params.py [hunk 3]
```diff
@@ -27,6 +27,11 @@ def load_arguments(self, _):
     with self.argument_context('monitor app-insights component update-tags') as c:
         c.argument('tags', tags_type)
 
+    with self.argument_context('monitor app-insights component billing') as c:
+        c.argument('stop_send_notification_when_hit_cap', options_list=['-s', '--stop'], arg_type=get_three_state_flag(),
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/_params.py]: stop_sending_notification_when_hitting_cap

Comment (MyronFanQiu) [reply src/application-insights/azext_applicationinsights/_params.py]: will change it.

File: src/application-insights/azext_applicationinsights/commands.py (hunk 1 above)
Comment (jsntcy) [reply src/application-insights/azext_applicationinsights/commands.py]: Make sense to me.

---
In reply to: [383775771](https://github.com/Azure/azure-cli-extensions/pull/1316#discussion_r383775771) [](ancestors = 383775771)

Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/commands.py]: typo: component

File: src/application-insights/azext_applicationinsights/commands.py [hunk 4]
```diff
@@ -60,6 +70,10 @@ def load_command_table(self, _):
...
+        g.custom_show_command('show', 'show_componet_billing')
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/commands.py]: type: component

File: src/application-insights/azext_applicationinsights/_client_factory.py [hunk 5]
```diff
@@ -16,28 +16,32 @@ def applicationinsights_data_plane_client(cli_ctx, _, subscription=None):
...
+    return applicationinsights_mgmt_plane_client(cli_ctx).components
 
 
-def cf_api_key(cli_ctx, _, subscription=None):
-    return applicationinsights_mgmt_plane_client(cli_ctx, _, subscription=subscription).api_keys
+def cf_component_billing(cli_ctx, _):
+    return applicationinsights_mgmt_plane_client(cli_ctx).component_current_billing_features
+
+
+def cf_api_key(cli_ctx, _):
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/_client_factory.py]: >_ [](start = 24, length = 1)

I'm curious why we need '_' as a method parameter.

File: src/application-insights/azext_applicationinsights/custom.py [hunk 6]
```diff
@@ -86,7 +86,7 @@ def show_api_key(client, application, resource_group_name, api_key=None):
     result = list(filter(lambda result: result.name == api_key, client.list(resource_group_name, application)))
     if len(result) == 1:
         return result[0]
-    elif len(result) > 1:
```
Comment (jsntcy) [inline src/application-insights/azext_applicationinsights/custom.py]: Looks like the logic doesn't change by changing elif to if

File: src/application-insights/azext_applicationinsights/_help.py (hunk 2 above)
Comment (jsntcy) [reply src/application-insights/azext_applicationinsights/_help.py]: simple but may be useful to users.

---
In reply to: [383775971](https://github.com/Azure/azure-cli-extensions/pull/1316#discussion_r383775971) [](ancestors = 383775971)

File: src/application-insights/azext_applicationinsights/_client_factory.py (hunk 5 above)
Comment (MyronFanQiu) [reply src/application-insights/azext_applicationinsights/_client_factory.py]: based on my shallow understanding right now, this is a limitation of azure_cli. `client` function must have at least two positional parameter. At least almost all of commands write like this. I'm not quite familiar with this part right now.

File: src/application-insights/azext_applicationinsights/custom.py (hunk 6 above)
Comment (MyronFanQiu) [reply src/application-insights/azext_applicationinsights/custom.py]: yes. The logic is same. But pylint ask us to use if instead of elif.

**Pull request merged**